controller_interface/include/controller_interface/chainable_controller_interface.hpp

```diff
@@ -0,0 +1,87 @@
+// Copyright (c) 2022, Stogl Robotics Consulting UG (haftungsbeschränkt)
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef CONTROLLER_INTERFACE__CHAINABLE_CONTROLLER_INTERFACE_HPP_
+#define CONTROLLER_INTERFACE__CHAINABLE_CONTROLLER_INTERFACE_HPP_
+
+#include <vector>
+
+#include "controller_interface/controller_interface.hpp"
+#include "controller_interface/visibility_control.h"
+#include "hardware_interface/handle.hpp"
+
+namespace controller_interface
+{
+/// Virtual class to implement when integrating a controller that can be preceded by another
+/// controllers.
+/**
+ * Specialization of ControllerInterface class to force implementation of methods specific for
+ * "chainable" controller, i.e., controller that can be preceded by an another controller, for
+ * example inner controller of an control cascade.
+ *
+ */
+class ChainableControllerInterface : public ControllerInterface
+{
+public:
+  CONTROLLER_INTERFACE_PUBLIC
+  ChainableControllerInterface();
+
+  CONTROLLER_INTERFACE_PUBLIC
+  virtual ~ChainableControllerInterface() = default;
+
+  CONTROLLER_INTERFACE_PUBLIC
+  bool is_chainable() const final;
+
+  CONTROLLER_INTERFACE_PUBLIC
+  std::vector<hardware_interface::CommandInterface> export_reference_interfaces() final;
+
+  CONTROLLER_INTERFACE_PUBLIC
+  bool set_chained_mode(bool chained_mode) final;
+
+  CONTROLLER_INTERFACE_PUBLIC
+  bool is_in_chained_mode() const final;
+
+protected:
+  /// Virtual method that each chainable controller should implement to export its chainable
+  /// interfaces.
+  /**
+   * Each chainable controller implements this methods where all input (command) interfaces are
+   * exported. The method has the same meaning as `export_command_interface` method from
+   * hardware_interface::SystemInterface or hardware_interface::ActuatorInterface.
+   *
+   * \returns list of CommandInterfaces that other controller can use as their outputs.
+   */
+  virtual std::vector<hardware_interface::CommandInterface> on_export_reference_interfaces() = 0;
```

Comment:
Hi @saikishor,

this is exactly something I am thinking about in the last days. Nevertheless, this feature will and have to come completely separated when we have finished chaining of command interfaces :)
My proposal, let's make this features finished, and then it will be much easier to get almost the same functionality with the state interfaces. Although there will be some complexities there to make this properly.

Reply:
@destogl I have few more questions regarding the implementation; This implementation with the CommandInterfaces if I understand right, it's when your controller is exposing some interfaces so other controllers can command it right?, like a position controller expecting joint commands from the higher level controllers?. Please correct me if I am wrong.

Regarding the naming convention of the interfaces, IMHO using `hardware_interfaces` to have the communication done between the controllers is strange, because `hardware_interfaces` as the name signifies  should be bound to the hardware, and for the controllers, we could have something like `controller_interfaces` or `software_interfaces`. What do you think about this?. My proposal is to move the part of the handle and other stuff to a new package and then reuse them in the current `hardware_interfaces` package and in the `controller_interfaces` or `software_interfaces`. If this is something that is not interesting atleast add a naming alias in the `hardware_interface` package. This makes easier for the developer to discriminate between them.

Reply:
> @destogl I have few more questions regarding the implementation; This implementation with the CommandInterfaces if I understand right, it's when your controller is exposing some interfaces so other controllers can command it right?, like a position controller expecting joint commands from the higher level controllers?. Please correct me if I am wrong.

This is correct!


> Regarding the naming convention of the interfaces, IMHO using `hardware_interfaces` to have the communication done between the controllers is strange, because `hardware_interfaces` as the name signifies  should be bound to the hardware, and for the controllers, we could have something like `controller_interfaces` or `software_interfaces`. What do you think about this?. My proposal is to move the part of the handle and other stuff to a new package and then reuse them in the current `hardware_interfaces` package and in the `controller_interfaces` or `software_interfaces`. If this is something that is not interesting atleast add a naming alias in the `hardware_interface` package. This makes easier for the developer to discriminate between them.


We are actually not using "hardware_interfaces". There is hardware interface package that where handles are defined. They are defined there because Thea are generally used by the hardware to provide access from controllers to it. Using the same type for handles for commanding hardware and providing inputs for controllers makes sense from functionality perspective because it is the same. Also this simplifies internal management of all interfaces which is also the same.

We could move this to another package, but I am not sure this would simplify things. It would result in an additional package with very little code and additional dependencies for all other packages.

I am also not sure this would add any clarity because developers does not have to do much with this types. They are used once or twice.

Can you try to make a concrete proposal how this changes would look like? Then we can find a common solution. I understand the issues you see, but I am not sure how to solve it without adding more complexity to the code base.

Reply:
I have an example: Typically I want to control the Center of Mass of my system. This is not really a hardware_interface but this is a reference quantity that I need to be controlled. For this I want the controller to take as an input the reference and the estimated value. 

Don't you think we could simply make :
`typedef CoMControllerInterface hardware_interface::CommandInterface`
and use the new defined type to make the controller interface clearer ?

Reply:
> I have an example: Typically I want to control the Center of Mass of my system. This is not really a hardware_interface but this is a reference quantity that I need to be controlled. For this I want the controller to take as an input the reference and the estimated value. 

Yes, this is one kind of use case I was also thinking too. I do get the idea behind the Command Interfaces.

> Don't you think we could simply make :
> `typedef CoMControllerInterface hardware_interface::CommandInterface`
> and use the new defined type to make the controller interface clearer ?

@olivier-stasse yes, ofcourse it makes it clear, but if we have 10 different types of references/inputs to your system, we don't want to define 10 different typedef lines. I don't think typedefs is a good idea in this case. My intention is that something that has symantic meaning would be interesting.

Reply:
Hello @destogl,

> Using the same type for handles for commanding hardware and providing inputs for controllers makes sense from functionality perspective because it is the same. Also this simplifies internal management of all interfaces which is also the same.

I completely agree with the idea behind it and also from the functionality perspective.

> We could move this to another package, but I am not sure this would simplify things. It would result in an additional package with very little code and additional dependencies for all other packages.

This is something we should think of. May be we can come up with a solution that requires less changes.

> I am also not sure this would add any clarity because developers does not have to do much with this types. They are used once or twice.

Well, maintaining symantic meaning is something important for a software like ROS. I think we should strive to maintain symantic meanings whenever we can.

> Can you try to make a concrete proposal how this changes would look like? Then we can find a common solution. I understand the issues you see, but I am not sure how to solve it without adding more complexity to the code base.

Sure, I can try to come back to you with a proposal. 

Thank you.